### PR TITLE
feat(server): user/item CRUD + library/genre discovery (#71, #73)

### LINKED ISSUE #71: Implement user ID resolution and item CRUD
Port `_get_user_id`, `get_item`, `update_item`, `prefetch_audio_items` (paginated). User-scoped endpoints: GET /Users/{id}/Items. POST /Items/{id} for updates.

### LINKED ISSUE #73: Implement library discovery and genre listing
Port `discover_libraries` (GET /Library/VirtualFolders, filter CollectionType=music) and `list_genres` (GET /MusicGenres?Recursive=true).

### PR DESCRIPTION
**Pull request by sydlexius**
## Summary
- Adds 6 typed response structs: `UserInfo`, `VirtualFolder`, `GenreItem`, `GenreResponse`, `AudioItemView`, `PrefetchResponse`
- Implements `get_user_id()` with `OnceCell` caching through `&self`
- Implements `get_item()` and `update_item()` for round-trip item updates
- Implements `prefetch_audio_items()` with paginated fetch (500 items/page, 3 termination conditions)
- Implements `discover_libraries()` (filters to music-type only)
- Implements `list_genres()` with case-insensitive sort via `sort_by_cached_key`
- `extract_audio_items()` uses borrowing deserialization and logs warnings for unparseable items

## Test plan
- [x] 27 unit tests: 7 type parsing tests, 2 extract_audio_items tests, plus all PR 1 tests
- [x] 11 UAT integration tests: user ID, prefetch, libraries, genres, get_item (both Emby + Jellyfin)
- [x] `cargo clippy -- -D warnings` passes

Stacked on #102. Closes #71, closes #73.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Enhanced media server integration: user lookup, fetch full item data, update items, batch prefetch of audio items with progress handling, discover music libraries, and list available genres; added audio-item extraction helper.
* **Tests**
  * Added integration tests for Emby/Jellyfin flows and expanded unit tests covering deserialization, extraction, and edge cases.

### CODE REVIEW

File: smpr/src/server/mod.rs [hunk 1]
```diff
@@ -174,6 +175,159 @@ impl MediaServerClient {
...
+}
+
+/// Extract (AudioItemView, Value) pairs from raw JSON item values.
+/// Logs a warning for items that fail to deserialize.
+pub fn extract_audio_items(items: &[Value]) -> Vec<(types::AudioItemView, Value)> {
+    items
+        .iter()
+        .filter_map(|v| {
+            match types::AudioItemView::deserialize(v) {
+                Ok(view) => Some((view, v.clone())),
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/mod.rs]: Fixed in de7927b — `extract_audio_items` now takes owned `Vec<Value>` and uses `into_iter()` to move values instead of cloning.

File: smpr/src/server/mod.rs [hunk 2]
```diff
@@ -174,6 +175,159 @@ impl MediaServerClient {
...
+        parent_id: Option<&str>,
+    ) -> Result<Vec<(types::AudioItemView, Value)>, MediaServerError> {
+        let mut fields = "Path,OfficialRating,AlbumArtist,Album,Genres".to_string();
+        if include_media_sources && self.server_type == ServerType::Emby {
+            fields.push_str(",MediaSources");
+        }
+        let uid = self.get_user_id()?;
+        let parent_filter = parent_id
+            .map(|id| format!("&ParentId={id}"))
+            .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/mod.rs]: Same as above — parent_id comes from `VirtualFolder.item_id`, a server-generated value.

File: smpr/src/server/tests/integration.rs [hunk 3]
```diff
@@ -40,3 +40,104 @@ fn detect_unreachable_returns_error() {
...
+    let Some(client) = jellyfin_client() else { return };
+    let uid = client.get_user_id().unwrap();
+    assert!(!uid.is_empty());
+}
+
+#[test]
+fn emby_prefetch_audio_items() {
+    let Some(client) = emby_client() else { return };
+    let items = client.prefetch_audio_items(false, None).unwrap();
+    assert!(!items.is_empty(), "expected at least one audio item");
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/tests/integration.rs]: Won't fix — UAT library has ~1500 items, takes ~13s. Acceptable for opt-in tests gated behind `SMPR_UAT_TEST=1`.

File: smpr/src/server/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/mod.rs]: Already fixed in de7927b — `extract_audio_items` now takes owned `Vec<Value>` with `into_iter()`.

File: smpr/src/server/tests/integration.rs [hunk 4]
```diff
@@ -40,3 +40,104 @@ fn detect_unreachable_returns_error() {
...
+        return None;
+    }
+    dotenvy::from_path(std::path::Path::new("../.env")).ok();
+    let key = std::env::var("UAT_JELLYFIN_API_KEY").ok()?;
+    Some(MediaServerClient::new(
+        "http://localhost:8097".to_string(),
+        key,
+        ServerType::Jellyfin,
+    ))
+}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/tests/integration.rs]: Won't fix — `let Some(...) else { return }` is idiomatic Rust for optional test skipping.

File: smpr/src/server/mod.rs [hunk 5]
```diff
@@ -174,6 +175,160 @@ impl MediaServerClient {
...
+            })?;
+        let resp: types::GenreResponse = serde_json::from_value(result)
+            .map_err(|e| MediaServerError::Parse(format!("MusicGenres: {e}")))?;
+        let mut names: Vec<String> = resp
+            .items
+            .into_iter()
+            .filter(|g| !g.name.is_empty())
+            .map(|g| g.name)
+            .collect();
+        names.sort_by_cached_key(|name| name.to_lowercase());
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/mod.rs]: Won't fix — genre display order among case-identical names is irrelevant. The sort ensures alphabetical grouping.

File: smpr/src/server/tests/integration.rs [hunk 6]
```diff
@@ -40,3 +40,104 @@ fn detect_unreachable_returns_error() {
...
+    let Some(client) = emby_client() else { return };
+    let genres = client.list_genres().unwrap();
+    assert!(!genres.is_empty(), "expected at least one genre");
+    // Verify sorted (case-insensitive)
+    let sorted: Vec<String> = {
+        let mut g = genres.clone();
+        g.sort_by(|a, b| a.to_lowercase().cmp(&b.to_lowercase()));
+        g
+    };
+    assert_eq!(genres, sorted);
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/tests/integration.rs]: Won't fix — same as above, sort stability among case-identical genres doesn't affect correctness.

File: smpr/src/server/tests/integration.rs [hunk 7]
```diff
@@ -40,3 +40,104 @@ fn detect_unreachable_returns_error() {
...
+    let Some(client) = emby_client() else { return };
+    let items = client.prefetch_audio_items(false, None).unwrap();
+    let (view, _) = &items[0];
+    let full_item = client.get_item(&view.id).unwrap();
+    assert!(full_item.get("Id").is_some());
+    assert!(full_item.get("Name").is_some());
+    // Verify round-trip body has more keys than the view
+    let keys = full_item.as_object().unwrap().len();
+    assert!(keys > 10, "expected full item body, got {keys} keys");
+}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/tests/integration.rs]: Acknowledged — Jellyfin get_item coverage noted as a minor gap in previous round. Emby is the primary target.

File: smpr/src/server/tests/integration.rs [hunk 8]
```diff
@@ -40,3 +40,110 @@ fn detect_unreachable_returns_error() {
...
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply smpr/src/server/tests/integration.rs]: Repeat of previous round — acknowledged, Jellyfin coverage is lighter by design.
